Address console warning for Image block inside Column block

kienstra opened the pull request:
## Description
There is a console warning that _sometimes_ appears when adding an Image block inside a Column block. 

We also saw this with the Image block inside custom block: https://github.com/ampproject/amp-wp/issues/2194#issue-437104605

This PR usually prevent this error locally. But the warning still appears sometimes with it. 

I'm not sure if it's the right approach, or if it has side-effects.

This is modeled after a few fixes for similar warnings, like https://github.com/WordPress/gutenberg/pull/5443.

Steps To Reproduce:

1. Create a post 
2. Add a "Columns block"
3. Add an "Image" block inside it
4. Click "Media Library" and select an image
5. A console warning appears:
```
react-dom.165d5c53.js:500 Warning: A component is changing an uncontrolled input of type number to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://fb.me/react-controlled-components
    in input
    in div (created by De)
    in div (created by De)
    in De
    in Unknown (created by WithInstanceId(Component))
    in WithInstanceId(Component) (created by t)
    in div (created by t)
    in div (created by t)
    in div (created by t)
    in t (created by ForwardRef(PanelBody))
    in ForwardRef(PanelBody) (created by t)
    in t (created by Context.Consumer)
    in se (created by InspectorControlsSlot)
    in InspectorControlsSlot
    in div
    in Unknown (created by n)
    in n (created by Context.Consumer)
    in WithSelect(Component)
    in div (created by t)
    in t (created by ForwardRef(PanelBody))
    in ForwardRef(PanelBody)
    in div (created by sn)
    in sn
    in div (created by sn)
    in sn (created by n)
    in div (created by n)
    in n (created by Context.Consumer)
    in Unknown (created by b)
    in b
    in t (created by Context.Consumer)
    in se (created by SidebarSlot)
    in SidebarSlot
    in div (created by t)
    in t
    in Unknown (created by n)
    in n (created by Context.Consumer)
    in WithViewportMatch(Component) (created by NavigateRegions(WithViewportMatch(Component)))
    in div (created by NavigateRegions(WithViewportMatch(Component)))
    in NavigateRegions(WithViewportMatch(Component)) (created by r)
    in r (created by Context.Consumer)
    in WithDispatch(NavigateRegions(WithViewportMatch(Component))) (created by n)
    in n (created by Context.Consumer)
    in WithSelect(WithDispatch(NavigateRegions(WithViewportMatch(Component)))) (created by t)
    in t (created by t)
    in div (created by t)
    in t (created by t)
    in t (created by t)
    in t (created by r)
    in r (created by Context.Consumer)
    in WithDispatch(t)
    in Unknown (created by Context.Consumer)
    in WithRegistryProvider(WithDispatch(t)) (created by t)
    in t (created by r)
    in r (created by Context.Consumer)
    in WithDispatch(t) (created by n)
    in n (created by Context.Consumer)
    in WithSelect(WithDispatch(t)) (created by t)
    in StrictMode (created by t)
    in t (created by n)
    in n (created by Context.Consumer)
    in WithSelect(t)
```

## How has this been tested?
* Performing the steps above, and observing that there usually isn't a console warning
* Still, this could use feedback, as I'm not sure this is the best approach

## Screenshots
Before PR:
![similar-warning](https://user-images.githubusercontent.com/4063887/56842158-5b51a980-6858-11e9-90a6-c880b440301a.gif)

## Types of changes

Bug fix: This sets a default value of `0` for `width` and `height`.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

kienstra committed:
Address a console warning for Image block inside Column block

There's a console warning:
Warning: A component is changing an uncontrolled
input of type number to be controlled.
I'm not sure this is the best approach,
or if it'll have side-effects.
But this is modeled after a few fixes for similar warnings,
like https://github.com/WordPress/gutenberg/pull/5443

jorgefilipecosta commented on packages/block-library/src/image/edit.js:
```diff
@@ -481,15 +481,15 @@ class ImageEdit extends Component {
 									className="block-library-image__dimensions__width"
 									label={ __( 'Width' ) }
-									value={ width !== undefined ? width : imageWidth }
+									value={ width ? width : imageWidth }
```
Hi @kienstra thank you for submitting this PR to address this issue. imageWidth/imageHeight also don't contain a default value:
```
https://github.com/WordPress/gutenberg//blob/079c11b0d011c228718388e829c9826e1f34183a/packages/block-library/src/image/edit.js#L576-L579
```

Is it possible that because of imageWidth/imageHeight also being undefined we still get an error? Should we also add a default value of 0 to this variable?

kienstra replied:
Hi @jorgefilipecosta,
Thanks, https://github.com/WordPress/gutenberg/pull/15239/commits/2655161018fcf6c1825b2dc0b8538046186cc2bd applies your suggestion. 

4 of the 5 times I tested it, the [console warning](https://github.com/WordPress/gutenberg/pull/15239#issue-274215441) didn't appear. When it did appear, maybe I forgot to do a hard refresh.

kienstra committed:
Set default values for imageWidth and imageHeight

As @jorgefilipecosta mentioned,
this might help in cases where the warning still appears:

A component is changing an uncontrolled input of type number to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component.

youknowriad commented on packages/block-library/src/image/edit.js:
```diff
@@ -375,8 +375,8 @@ class ImageEdit extends Component {
-			width,
-			height,
+			width = 0,
+			height = 0,
```
For me undefined width/height is different in meaning from a width=0 and height=0 and the checks that were changed line 484 illustrates this. 

I think a safer fix would be to avoid adding default values here and instead treat the symptom (add a check to the inputs triggering the warning) `width || ""`, `height || ""` (I don't think we want to show 0 in these inputs when the value is undefined)

swissspidy replied:
Hmm that would work for me too. No strong opinion though.

kienstra replied:
Hi @youknowriad, thanks for your suggestion.

How does https://github.com/WordPress/gutenberg/pull/15239/commits/282e052b40bd61700c4a5929d0a83989123678f7 look? I'm not sure if it captures your suggestion.

There's a failed [E2E test](https://travis-ci.com/WordPress/gutenberg/jobs/199450362) that I could fix.

youknowriad replied:
Yep, that's exactly what I suggested, thanks.

The test failure is unrelated to this PR (restarted and passed) and is being worked on separately.

kienstra committed:
Remove the default values, and instead use width || ''

As Riad suggested,
0 is different from undefined,
and it's a little strange to force values to 0 instead of undefined.

youknowriad commented on packages/block-library/src/image/edit.js:
```diff
@@ -481,15 +481,15 @@ class ImageEdit extends Component {
 									className="block-library-image__dimensions__width"
 									label={ __( 'Width' ) }
-									value={ width !== undefined ? width : imageWidth }
+									value={ width !== undefined ? width || '' : imageWidth }
```
Something I'm wondering here. If `width` is not undefined this means there's no need for `|| ''` in the first part of the ternary right? Is the warning triggered by the first part or the second part (without this PR)?

kienstra replied:
Hi @youknowriad, 
Actually, when I tested this again, it looks like this ternary doesn't prevent the console warning.

https://github.com/WordPress/gutenberg/pull/15239/commits/282e052b40bd61700c4a5929d0a83989123678f7, which adds this ternary, looks to be a regression where the console warning appears again.

You made a good point in https://github.com/WordPress/gutenberg/pull/15239#discussion_r281590329 about the default value of `0` being strange. But I'm not sure how to prevent this console warning otherwise.

Thanks!

swissspidy commented:
I just tested the following and it seems to work:

```diff
diff --git packages/block-library/src/image/edit.js packages/block-library/src/image/edit.js
index 91eeb5a40..645248bc7 100644
--- packages/block-library/src/image/edit.js
+++ packages/block-library/src/image/edit.js
@@ -482,7 +482,7 @@ class ImageEdit extends Component {
 									type="number"
 									className="block-library-image__dimensions__width"
 									label={ __( 'Width' ) }
-									value={ width !== undefined ? width : imageWidth }
+									value={ width || imageWidth || '' }
 									min={ 1 }
 									onChange={ this.updateWidth }
 								/>
@@ -490,7 +490,7 @@ class ImageEdit extends Component {
 									type="number"
 									className="block-library-image__dimensions__height"
 									label={ __( 'Height' ) }
-									value={ height !== undefined ? height : imageHeight }
+									value={ height || imageHeight || '' }
 									min={ 1 }
 									onChange={ this.updateHeight }
 								/>

```

kienstra commented:
Thanks, @swissspidy! Feel free to push that to this branch if you'd like. Otherwise, I'll push it and credit you.

swissspidy commented:
@kienstra If that change works for you, feel free to #shipit 🙂

kienstra committed:
Apply @swisspidy's commit to prevent console warning

This uses the || operator entirely,
instead of a ternary.
Props @swisspidy

kienstra commented:
Thanks, @swissspidy! I just committed your change. Though it looks like the build failed, with a failed [E2E test](https://travis-ci.com/WordPress/gutenberg/jobs/200222674).

kienstra commented:
Would you be able to restart the build, to see if it'll pass? Thanks!

youknowriad reviewed (approved):
LGTM, Thanks for the fix
